python-sdk/src/astro/utils/table.py

```diff
@@ -1,97 +1,122 @@
+from astro.utils.typing_compat import Context
 
 
-def _pull_first_table_from_parameters(parameters: dict) -> Optional[BaseTable]:
+def _have_same_conn_id(tables: list[BaseTable]) -> bool:
     """
-    When trying to "magically" determine the context of a decorator, we will try to find the first table.
-    This function attempts this by checking parameters
+    Check to see if all tables belong to same conn_id. Otherwise, this can go wrong for cases
+    1. When we have tables from different DBs.
+    2. When we have tables from different conn_id, since they can be configured with different database/schema etc.
 
-    :param parameters: a user-defined dictionary of parameters
-    :return: the first parameter of type BaseTable, if any. Return None otherwise.
+    :param tables: tables to check for conn_id field
+    :return: True if all tables have the same conn id, and False if not.
+    """
+    return len(tables) == 1 or len({table.conn_id for table in tables}) == 1
+
+
+def _find_first_table_from_op_args(
+    op_args: tuple, context: Context
+) -> BaseTable | None:
+    """
+    Read op_args and extract the tables.
+
+    :param op_args: user-defined operator's args
+    :param context: the context to use for resolving XComArgs
+    :return: first valid table found in op_args.
     """
-    first_table = None
-    params_of_table_type = [
-        param for param in parameters.values() if isinstance(param, BaseTable)
+    args = [
+        arg.resolve(context) if isinstance(arg, XComArg) else arg for arg in op_args
     ]
-    if (
-        len(params_of_table_type) == 1
-        or len({param.conn_id for param in params_of_table_type}) == 1
-    ):
-        first_table = params_of_table_type[0]
-    return first_table
+    tables = [arg for arg in args if isinstance(arg, BaseTable)]
 
+    if _have_same_conn_id(tables):
+        return tables[0]
+    return None
 
-def _pull_first_table_from_op_kwargs(
-    op_kwargs: dict, python_callable: Callable
-) -> Optional[BaseTable]:
+
+def _find_first_table_from_op_kwargs(
+    op_kwargs: dict, python_callable: Callable, context: Context
+) -> BaseTable | None:
     """
-    When trying to "magically" determine the context of a decorator, we will try to find the first table.
-    This function attempts this by checking op_kwargs
+    Read op_kwargs and extract the tables.
 
     :param op_kwargs: user-defined operator's kwargs
-    :return: the first kwarg declared in the decorated method which is an
-        instance of BaseTable. Return None if non-existent
+    :param context: the context to use for resolving XComArgs
+    :return: first valid table found in op_kwargs.
     """
-    first_table = None
-    kwargs_of_table_type = [
-        op_kwargs[kwarg.name]
+    kwargs = [
+        op_kwargs[kwarg.name].resolve(context)
+        if isinstance(op_kwargs[kwarg.name], XComArg)
+        else op_kwargs[kwarg.name]
         for kwarg in inspect.signature(python_callable).parameters.values()
```

Comment:
I have the feeling we're trying to accomplish a lot in a single list comprehension - would it be worth to break this logic in (a) separate function(s)?

Reply:
🤔 not sure.. It collects all kwargs from the python callable signature and if they are XcomArg they will be resolved first.